xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+	// part of this response that are in state `started` or `requested`,
+	// this includes:
+	//   - setting the watch state to watchstateReceived
+	//   - stopping the expiry timer, if one exists
+	for _, name := range names {
+		rs, ok := typeState.subscribedResources[name]
+		if !ok {
+			s.logger.Warningf("ADS stream received a response for resource %q, but no state exists for it", name)
+			continue
+		}
+		if ws := rs.State; ws == ResourceWatchStateStarted || ws == ResourceWatchStateRequested {
+			rs.State = ResourceWatchStateReceived
+			if rs.ExpiryTimer != nil {
+				rs.ExpiryTimer.Stop()
+				rs.ExpiryTimer = nil
+			}
+		}
+	}
+
+	// Send an ACK or NACK.
+	subscribedResourceNames := resourceNames(typeState.subscribedResources)
+	if nackErr != nil {
+		s.logger.Warningf("Sending NACK for resource type: %q, version: %q, nonce: %q, reason: %v", url, version, nonce, nackErr)
+		s.sendMessageLocked(stream, subscribedResourceNames, url, previousVersion, nonce, nackErr)
+	} else {
+		if s.logger.V(2) {
+			s.logger.Infof("Sending ACK for resource type: %q, version: %q, nonce: %q", url, version, nonce)
+		}
+		s.sendMessageLocked(stream, subscribedResourceNames, url, version, nonce, nil)
+	}
+}
+
+func (s *Stream) onError(err error, msgReceived bool) {
+	s.mu.Lock()
+	// For resources that been requested but not yet responded to by the
+	// management server, stop the resource timers and reset the watch state to
+	// watchStateStarted.
+	for _, state := range s.resourceTypeState {
+		for _, rs := range state.subscribedResources {
+			if rs.State != ResourceWatchStateRequested {
+				continue
+			}
+			if rs.ExpiryTimer != nil {
+				rs.ExpiryTimer.Stop()
+				rs.ExpiryTimer = nil
+			}
+			rs.State = ResourceWatchStateStarted
+		}
+	}
+
+	// Forward the error to the channel.
+	if msgReceived {
+		err = xdsresource.NewErrorf(xdsresource.ErrTypeStreamFailedAfterRecv, err.Error())
+	}
+	s.mu.Unlock()
+
+	s.eventHandler.OnADSStreamError(err)
+}
+
+// Caller must hold s.mu.
+func (s *Stream) startWatchTimersLocked(typ xdsresource.Type, names []string) {
+	typeState := s.resourceTypeState[typ]
+	for _, name := range names {
+		resourceState, ok := typeState.subscribedResources[name]
+		if !ok {
+			continue
+		}
+		if resourceState.State != ResourceWatchStateStarted {
+			continue
+		}
+		resourceState.State = ResourceWatchStateRequested
+
+		rs := resourceState
+		resourceState.ExpiryTimer = time.AfterFunc(s.watchExpiryTimeout, func() {
+			s.mu.Lock()
+			rs.State = ResourceWatchStateTimeout
+			rs.ExpiryTimer = nil
+			s.mu.Unlock()
+			s.eventHandler.OnADSWatchExpiry(typ, name)
+		})
+	}
+}
+
+func resourceNames(m map[string]*ResourceWatchState) []string {
+	ret := make([]string, 0, len(m))
+	for i := range m {
+		ret = append(ret, i)
+	}
+	return ret
+}
+
+// TriggerResourceNotFoundForTesting triggers a resource not found event for the
+// given resource type and name.  This is intended for testing purposes only, to
```

Comment:
Nit: drop double space. "Should only be called from tests".

Reply:
Intended is less strong than should only. But I don't really feel too strongly about this.